src/engine/Service.cpp

```diff
@@ -0,0 +1,222 @@
+  return parsedServiceClause_.visibleVariables_.size();
+}
+
+// ____________________________________________________________________________
+std::vector<size_t> Service::resultSortedOn() const { return {}; }
+
+// ____________________________________________________________________________
+VariableToColumnMap Service::computeVariableToColumnMap() const {
+  VariableToColumnMap map;
+  const auto& visibleVariables = parsedServiceClause_.visibleVariables_;
+  for (size_t i = 0; i < visibleVariables.size(); i++) {
+    map[visibleVariables[i]] = i;
+  }
+  return map;
+}
+
+// ____________________________________________________________________________
+float Service::getMultiplicity([[maybe_unused]] size_t col) {
+  // TODO: For now, we don't have any information about the multiplicities at
+  // query planning time, so we just return `1` for each column.
+  return 1;
+}
+
+// ____________________________________________________________________________
+size_t Service::getSizeEstimate() {
+  // TODO: For now, we don't have any information about the result size at
+  // query planning time, so we just return `100'000`.
+  return 100'000;
+}
+
+// ____________________________________________________________________________
+size_t Service::getCostEstimate() {
+  // TODO: For now, we don't have any information about the cost at query
+  // planning time, so we just return ten times the estimated size.
+  return 10 * getSizeEstimate();
+}
+
+// ____________________________________________________________________________
+void Service::computeResult(ResultTable* result) {
+  using Url = ad_utility::httpUtils::Url;
+
+  // Get the URL of the SPARQL endpoint (we can only do fixed IRIs at the
+  // moment).
+  if (std::holds_alternative<Variable>(parsedServiceClause_.varOrIri_)) {
+    throw std::runtime_error(
+        "SERVICE with variable for endpoint not yet supported by QLever");
+  }
+  auto iri = std::get<GraphTerm>(parsedServiceClause_.varOrIri_);
+  // According to the grammar, this can only be of type `Iri`. See the comment
+  // in `Service.h` for `varOrIri_`.
+  AD_CHECK(std::holds_alternative<Iri>(iri));
+  std::string_view serviceIriString = std::get<Iri>(iri).iri();
+  AD_CHECK(serviceIriString.starts_with("<") &&
+           serviceIriString.ends_with(">"));
+  serviceIriString.remove_prefix(1);
+  serviceIriString.remove_suffix(1);
+  Url serviceUrl{serviceIriString};
+
+  // Construct the query to be sent to the SPARQL endpoint.
+  std::string variablesForSelectClause =
+      absl::StrJoin(parsedServiceClause_.visibleVariables_, " ",
+                    [](std::string* out, const Variable& variable) {
+                      out->append(variable.name());
+                    });
+  std::string serviceQuery = absl::StrCat(
+      parsedServiceClause_.prologue_, "\nSELECT ", variablesForSelectClause,
+      " WHERE ", parsedServiceClause_.graphPatternAsString_);
+  LOG(INFO) << "Sending SERVICE query to remote endpoint "
```

Comment:
I would prefer to see basic information about each operation in the log. It would be great though to have a level between INFO and DEBUG. Like "INFO minimal" and "INFO normal".